cmd/ipfs/daemon.go

```diff
@@ -459,6 +460,16 @@ func daemonFunc(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment
 
 	printSwarmAddrs(node)
 
+	if node.PrivateKey.Type() == p2pcrypto.RSA {
+		fmt.Print(`
+You are using an RSA Peer ID, thoses are deprecated and support will be removed sometime second half of 2023.
```

Comment:
Was this discussed at an IPFS implementers sync with some consensus? This seems like you'd be breaking every kubo node currently in existence (including from the current release) due to bootstrap nodes having RSA keys.

In exchange, you save a little bit of CPU time? IMO this seems like a really bad tradeoff, so I'd like to be able to review the discussion around it. It will also be a helpful resource for people this June who start asking why things have silently stopped working.

For comparison, when we dropped support for SECIO there was a more pressing reason, a smoother upgrade path and more communication for why we were doing so.

Reply:
I've talked with @lidel and @marten-seemann about adding a disclaimer.

It didn't occured to me that deprecation should be discussed at the IPFS implementer sync given how small the fraction of the network that represent.
I also suspect that the share of RSA nodes is higher in non diallable node somehow, because bootrstrap nodes and hydras see notable CPU usage reduction when using openssl instead of the gostd for crypto, this 3% ain't significant.

I'll change the message to be a time indefinite warning instead of a deprecation before bringing this up to the IPFS implementers sync.

Reply:
> This seems like you'd be breaking every kubo node currently in existence (including from the current release) due to bootstrap nodes having RSA keys.

Ah ... yes, I guess it would be easy to make new bootstrap nodes by then.

Reply:
Is libp2p, or go-libp2p planning on dropping support for RSA keys, or is this an IPFS-specific thing being proposed? Given that kubo supports keys like secp256k1 and ecdsa that are almost unused on the IPFS network, seems weird to drop RSA just because the handshakes are a bit more expensive. If libp2p is on board with this (and has low consumer pushback) it seems to me easier to get on board from the IPFS side of things.

> I also suspect that the share of RSA nodes is higher in non dialable node somehow

This seems worth checking into, I would also leverage the existing metrics on peer distributions collected within PL infra (e.g. that looks at peerIDs and user agents). The data is already being gathered so it's likely not too bad to just add another check.

> given how small the fraction of the network that represent.

maybe ... per above. Technically, the bootstrappers all being RSA effectively means either implementing hacky hydra-like behavior on the existing bootstrappers or having a network fork (even for Ed25519 key holders). Both of these are a pain. Which begs the question if this is currently worth the pain.

Reply:
> I also suspect that the share of RSA nodes is higher in non diallable node somehow, because bootrstrap nodes and hydras see notable CPU usage reduction when using openssl instead of the gostd for crypto, this 3% ain't significant.

I've heard this claim a few times, but I've never seen actual numbers. Every time I grabbed a pprof, signature verification didin't even show up.

> Is libp2p, or go-libp2p planning on dropping support for RSA keys, or is this an IPFS-specific thing being proposed? Given that kubo supports keys like secp256k1 and ecdsa that are almost unused on the IPFS network, seems weird to drop RSA just because the handshakes are a bit more expensive. If libp2p is on board with this (and has low consumer pushback) it seems to me easier to get on board from the IPFS side of things.

No, we'll keep supporting RSA keys. From my benchmarks, there's no problem with them, they're not even much slower than Ed25519 keys CPU wise. They're big, so you'll be hurting your network performance a bit.

Reply:
> EDIT: Bootstrap nodes high RSA cost is self inflicted.

I see, that makes more sense. _Signing_ is expensive, not verifying: https://github.com/libp2p/go-libp2p/pull/1686#issuecomment-1347546446.